### PR TITLE
bpo-32025: Add time.thread_time()

### PR DESCRIPTION
**Pull request by pitrou**
https://bugs.python.org/issue32025

### CODE REVIEW

File: Doc/library/time.rst [hunk 1]
```diff
@@ -603,6 +604,31 @@ Functions
...
+      single: processor time
+      single: benchmarking
+
+   Return the value (in fractional seconds) of the sum of the system and user
+   CPU time of the current thread.  It does not include time elapsed during
+   sleep.  It is thread-specific by definition.  The reference point of the
+   returned value is undefined, so that only the difference between the results
+   of consecutive calls in the same thread is valid.
+
+   Availability:  Unix systems supporting CLOCK_THREAD_CPUTIME_ID, Windows.
```
Comment (vstinner) [inline Doc/library/time.rst]: I propose:

Availability: Windows, Linux, Unix systems supporting CLOCK_THREAD_CPUTIME_ID.

File: Modules/timemodule.c [hunk 2]
```diff
@@ -1258,6 +1258,112 @@ Process time for profiling as nanoseconds:\n\
...
+#elif defined(HAVE_CLOCK_GETTIME) && defined(CLOCK_PROCESS_CPUTIME_ID)
+#define HAVE_THREAD_TIME
+static int
+_PyTime_GetThreadTimeWithInfo(_PyTime_t *tp, _Py_clock_info_t *info)
+{
+    struct timespec ts;
+    const clockid_t clk_id = CLOCK_THREAD_CPUTIME_ID;
+    const char *function = "clock_gettime(CLOCK_THREAD_CPUTIME_ID)";
+
+    if (clock_gettime(clk_id, &ts) == 0) {
```
Comment (vstinner) [inline Modules/timemodule.c]: I suggest to invert the test to return early on error.

**Review by vstinner — APPROVED** [review]
LGTM. I just have two minor comments.

File: Misc/NEWS.d/next/Library/2017-11-15-20-03-45.bpo-32025.lnIKYT.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Add time.thread_time()
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2017-11-15-20-03-45.bpo-32025.lnIKYT.rst]: And time.thread_time_ns().

File: Modules/timemodule.c [hunk 4]
```diff
@@ -1258,6 +1258,112 @@ Process time for profiling as nanoseconds:\n\
...
+    if (clock_gettime(clk_id, &ts) != 0) {
+        PyErr_SetFromErrno(PyExc_OSError);
+        return -1;
+    }
+    if (info) {
+        struct timespec res;
+        info->implementation = function;
+        info->monotonic = 1;
+        info->adjustable = 0;
+        if (clock_getres(clk_id, &res)) {
```
Comment (serhiy-storchaka) [inline Modules/timemodule.c:1319]: `(clock_gettime(clk_id, &ts) != 0)` but `(clock_getres(clk_id, &res))`.

File: Lib/test/test_time.py [hunk 5]
```diff
@@ -486,10 +496,57 @@ def test_process_time(self):
...
+        stop = time.thread_time()
+        # use 20 ms because thread_time() has usually a resolution of 15 ms
+        # on Windows
+        self.assertLess(stop - start, 0.020)
+
+        # thread_time() should include CPU time spent in current thread...
+        start = time.thread_time()
+        busy_wait(0.100)
+        stop = time.thread_time()
+        self.assertGreaterEqual(stop - start, 0.020)  # machine busy?
```
Comment (serhiy-storchaka) [inline Lib/test/test_time.py:536]: Shouldn't it be `0.100 - 0.020` or maybe just `0.100`?

Comment (vstinner) [reply Lib/test/test_time.py:536]: Please don't be too strict, or the test will fail randomly on grumpy buildbots.

Comment (pitrou) [reply Lib/test/test_time.py:536]: The whole CPU time in those 0.1 seconds was not necessarily allocated to the current process / thread.

File: Modules/timemodule.c (hunk 4 above)
Comment (pitrou) [reply Modules/timemodule.c:1319]: This is really a copy-paste job.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**